### DISCUSSION TITLE
Default Terraform depName should include mode path

### DISCUSSION THREAD

**Discussion by twang817** [renovatebot/renovate#24822] [Suggest an Idea]
### What would you like Renovate to be able to do?

Currently, the Terraform manager identifies a a Github-based module source package by its repository URL:

```terraform
module "module1" {
  source = "git@github.com:owner/repo.git//modules/module1?ref=v1.2.3"
}
```

In this case, the code currently sets the packageName and depName both to `github.com/owner/repo`.

However, it is somewhat common practice for an organization to create a monorepo containing their terraform modules:

```
my-org-tf-modules:
.
└── modules
    ├── module1
    │   └── main.tf
    └── module2
        └── main.tf
```

In the event that your use both `module1` and `module2` from the same monorepo, this has the effect that renovate will create a single PR to update both modules.  This has come up a few times in discussions ([here](https://github.com/renovatebot/renovate/discussions/9557), [here](https://github.com/renovatebot/renovate/discussions/12331), and [here](https://github.com/renovatebot/renovate/discussions/21615)).

So far, it seems like the recommended approach is to abandon the Terraform manager and use a Regex Manager.

I believe that the Terraform manager should, by default, set the depName of each module it encounters to include the module path as well.  In our example above, packageName would remain unchanged while depName would be set to `github.com/owner/repo//modules/module1`.

This has the intended effect that any updates to modules sharing the same Git repository will be split into multiple PRs.

### If you have any ideas on how this should be implemented, please tell us here.

I've been able to get this working by changing the following:

```diff
diff --git a/lib/modules/manager/terraform/extractors/others/modules.ts b/lib/modules/manager/terraform/extractors/others/modules.ts
index 2c135d995..6db9b6412 100644
--- a/lib/modules/manager/terraform/extractors/others/modules.ts
+++ b/lib/modules/manager/terraform/extractors/others/modules.ts
@@ -10,7 +10,7 @@ import { DependencyExtractor } from '../../base';
 import type { TerraformDefinitionFile } from '../../hcl/types';

 export const githubRefMatchRegex = regEx(
-  /github\.com([/:])(?<project>[^/]+\/[a-z0-9-_.]+).*\?(depth=\d+&)?ref=(?<tag>.*?)(&depth=\d+)?$/i
+  /github\.com([/:])(?<project>[^/]+\/[a-z0-9-_.]+)(?<module_path>.*)\?(depth=\d+&)?ref=(?<tag>.*?)(&depth=\d+)?$/i
 );
 export const bitbucketRefMatchRegex = regEx(
   /(?:git::)?(?<url>(?:http|https|ssh)?(?::\/\/)?(?:.*@)?(?<path>bitbucket\.org\/(?<workspace>.*)\/(?<project>.*).git\/?(?<subfolder>.*)))\?(depth=\d+&)?ref=(?<tag>.*?)(&depth=\d+)?$/
@@ -67,7 +67,7 @@ export class ModuleExtractor extends DependencyExtractor {
         ''
       );
       dep.depType = 'module';
-      dep.depName = 'github.com/' + dep.packageName;
+      dep.depName = 'github.com/' + dep.packageName + githubRefMatch.groups.module_path;
       dep.currentValue = githubRefMatch.groups.tag;
       dep.datasource = GithubTagsDatasource.id;
     } else if (bitbucketRefMatch?.groups) {
```

This adds to the github regex to grab the module_path as a capture group, and uses it in the depName.  I do not know the other repository syntaxes well enough to make an equivalent change in bitbucket, gitlab, etc.  Nor am I confident in my understanding of the rest of renovate to understand what implications this might have to the rest of the code.

### Is this a feature you are interested in implementing yourself?

Yes

**Comment by rarkins** [comment]
Alternatively can just the module name be used as depName?

**Comment by twang817** [comment]
> Alternatively can just the module name be used as depName?

It is possible to source two completely different modules with the same path component from two different repos.  Consider:

```terraform
module "module1" {
  source = "git@github.com:owner/repo1.git//module?ref=v1.2.3"
}

module "module2" {
  source = "git@github.com:owner/repo2.git//module?ref=v1.2.3"
}
```

I think the depName should continue to include the github repo url.

**Comment by rarkins** [comment]
What about the strings "module1" and "module2" in your example?

**Comment by twang817** [comment]
Using the module logical names ("module1" and "module2" in the example) is probably more correct.

I believe Terraform will create a directory structure like this when initializing the plan:

```
.terraform/
└── modules/
    ├── module1/
    │   └── module/
    │       └── main.tf
    └── module2/
        └── module/
            └── main.tf
```

E.g., it uses the module's logical name on the filesystem.  This allows interesting use-cases like:

```
module "module1" {
  source = "git@github.com:owner/repo1.git//module?ref=v1.2.3"
}

module "module2" {
  source = "git@github.com:owner/repo1.git//module?ref=v4.5.6"
}
```

Using repo/module_path in this situation would revert back to current behavior, as both of these modules above would be treated as a single dependency.  Instead, using module1 and module2 would continue to provide the correct behavior.
